src/with_clause_parser.c

```diff
@@ -161,12 +162,37 @@ parse_arg(WithClauseDefinition arg, DefElem *def)
 
 	Assert(OidIsValid(in_fn));
 
+	/*
+	 * We could use InputFunctionCallSafe() here but this is just supported
+	 * for PG16 and later, so we opt for checking if the failure is what we
+	 * expected and re-throwing the error otherwise.
+	 */
 	PG_TRY();
 	{
 		val = OidInputFunctionCall(in_fn, value, typIOParam, -1);
 	}
 	PG_CATCH();
 	{
+		MemoryContext ecxt = MemoryContextSwitchTo(ccxt);
+		ErrorData *edata = CopyErrorData();
```

Comment:
Why do we need to copy the error data here? It doesn't seem like ErrorData is used after FlushErrorState anyway. ErrorData is used only to check the sqlerrcode, which happens before FlushErrorState() so it should be valid.